### PR TITLE
Simplify theme settings

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
As the title states, the goal here is to simplify the theme settings by removing the setting of theme for dark/light system preference. Let me explain my logic behind this:

If you need to open the settings menu to change the used theme, then, why would you care about changing your preferred dark/light theme and not just update the theme directly?

As a secondary benefit, it simplifies the JS code handling themes.

You can test it [here](https://rustdoc.crud.net/imperio/simplify-theme-settings/foo/index.html).

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Fixed eslint errors.

**Comment by Urgau** [comment]
> If you need to open the settings menu to change the used theme, then, why would you care about changing your preferred dark/light theme and not just update the theme directly?

You might have setup your system so that the system theme (and so your browser theme) follows the day/night cycle of your location and by that means to might want to set day=light and night=ayu. And if I understand correctly this PR this wouldn't be possible anymore, right ?

**Comment by GuillaumeGomez** [comment]
You could still follow the day cycle but not pick which dark/light theme is picked indeed.

**Comment by Urgau** [comment]
> You could still follow the day cycle but not pick which dark/light theme is picked indeed.

🤔 That not great. I could see people wanting this setup (there probably are people currently having it). Would be great to at least have a alternative if this PR lands.

**Comment by GuillaumeGomez** [comment]
I wonder how many people use this setup. Another thing I was planning to do was to switch the default dark theme to ayu. Might be interesting to see how many people have a day/light setting with ayu/dark as default dark theme.

**Comment by apiraino** [comment]
> You might have setup your system so that the system theme (and so your browser theme) follows the day/night cycle of your location

I was also confused by this at the beginning. Websites provide a dual setting, tipically the :crescent_moon:  / :sun_with_face: icon with no explicit option for a "system default". 

On MDN, the documentation for the [`prefer-color-scheme`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme) says:
- `light`: means "use dark theme" *or* no active user preference
- `dark` means "use dark theme"

So I think that a "system setting" will be automatically used when flipping the switch to `light`. Therefore no explicit "use system setting" should be needed.

Makes sense?

**Comment by GuillaumeGomez** [comment]
I'm not a big fan of implicit. If we want to allow users to switch back to "system preference", we should maybe just a new entry in the theme list.

**Comment by steffahn** [comment]
I like the idea of having "system preference" simply as a 4th option.

The most minimalistic additional configuration option to serve most use cases (in my opinion), given that we have one light theme and two dark themes, is to simply add a second option to choose the preferred dark theme. In order to make this choice minimally confusing, I would suggest that it only becomes visible if:
* the "system preference" option is chosen, ***and***
* the system settings *currently* are set to dark-themed

this way, every change of preference will always become immediately visible, there’s no hypothetical/dead options that would only take effect under certain conditions (w.r.t. system preferences) that you might never encounter, and there’s no weird options of calling dark themes “the preferred light theme” or vice-versa.

Here’s some mock-up screenshots of all possible states:

## ayu
Screenshot_20220702_165638

## dark
Screenshot_20220702_165649

## light
Screenshot_20220702_165709

## system preference (system theme is currently light-themed), preferred dark theme is saved but not displayed
Screenshot_20220702_165425

## system preference (system theme is currently dark-themed), preferred dark theme is ayu
Screenshot_20220702_165310

## system preference (system theme is currently dark-themed), preferred dark theme is dark
Screenshot_20220702_165133

**Comment by apiraino** [comment]
(not an expert, just my .2 cents)

Although from a UX point of view expliciting the option "use system theme" would be better, why can't I think of websites allowing such an option? :shrug: This is why I thought that a "use system default" was redundant and my understanding of the documentation for the CSS directive seems to confirm that.

The second menu option for the theme selection:
grafik

is more or less that we have now on rustdoc stable and imho it's a bit confusing, so in this regard I'm with @GuillaumeGomez in trying to reduce the actionable knobs.

**Comment by steffahn** [comment]
> The reason I don't like the four-option mockup is that it precludes adding a second light theme.

Within the context of what I presented above, a 2-light-themes + 2-dark-themes situation would be modeled without any problem; the idea that the “preferred dark theme” choice is *only* visible while “system preference” is active ***and*** the system preference is currently set to dark-themed (in order to *avoid* confusion from options that don’t have any immediate effect) translates into the same idea for “preferred light theme” option visibility, so you only see one-at-a-time. I.e. it looks like this

## system preference (system theme is currently light-themed), preferred light theme is light, preferred dark theme is saved but not displayed
Screenshot_20220702_201738

## system preference (system theme is currently light-themed), preferred light theme is rust, preferred dark theme is saved but not displayed
Screenshot_20220702_202015
(I only changed the background color, this is not a real theme)

## system preference (system theme is currently dark-themed), preferred dark theme is ayu, preferred light theme is saved but not displayed
Screenshot_20220702_201618

## system preference (system theme is currently dark-themed), preferred dark theme is dark, preferred light theme is saved but not displayed
Screenshot_20220702_201554

## some theme selected unconditionally (e.g. dark theme)
Screenshot_20220702_202326

To be clear, the choice of label for “system preference” I don’t exactly care about as long as it’s some easy-to-understand label.

---

One important consequence of this approach is: Whatever you click on is **always** what you get, immediately; and you can always choose any theme.

* You click on any theme in the top row and immediately get it. Single step, trivial discoverability. Especially considering that users typically start out on “system preference” by default, it’s important that *all themes* are visibly *somewhere*, otherwise you might not even know about dark themes existing in the first place ***despite looking into the options***.
* If you click on “system preference”, you will get a light theme with light system preference, and a dark theme with dark system preference, because the option to select light “preferred dark theme”s or dark “preferred light theme”s is gone
* There *is* e.g. two ways of choosing a dark theme while in dark-themed system preference mode, but both get you the right theme immediately, the difference only matters if your system theme changes, and you can easily fix the problem after the fact.
* Admitted, configuring a setup for using the system theme with custom choices for both dark and light theme *ahead of time* is no longer possible, but it should rarely be really necessary anyways. Someone who uses changing system themes on a day-night-cycle will have to choose their preferred day theme once *while it’s currently daytime* and they will have to choose their preferred night theme once *while it’s currently nighttime*, but the benefits you get for this limitation is a more intuitive user experience, because at least it’s always: what you click on is what you see and what you get.

By contrast (to the last point about getting what you click on), assuming I’m not misinterpreting [your mock-up](https://user-images.githubusercontent.com/1593513/177011290-73541d6a-8675-475e-81fa-240ad2f2650c.png), what your suggesting is: that the “preferred dark theme” option ***is*** visible while “use system theme” is on, and the system theme is light-themed, (which means by the way, that there’s a selection for dark themes that *can seem kind-of broken* to users, because it doesn’t have any visible immediate effects), and that the *real* option for selecting a dark theme ***isn't*** visible while “use system theme” is on. Essentially users would have to
* ignore the visible button labeled “dark” (in the “preferred dark theme” section) _… but hey, I **do** prefer dark themes!?_
* press the button labeled “use system theme” in order to *not* use the system theme (I know it’s an on/off switch, but that might not be obvious), then (it’s a 2-step-process!) click the *newly appearing* option to use a button labeled “dark” that looks almost identical to the button labeled “dark” that you *previously had to ignore*.

**Comment by steffahn** [comment]
> To be clear, the choice of label for “system preference” I don’t exactly care about as long as it’s some easy-to-understand label.

For example, the label could also say “sync with system” (like the GitHub preferences). Furthermore the title of the “preferred … theme” sections could also more clearly relate to this setting by saying something like “Dark theme for “sync with system””.
Screenshot_20220702_211323

**Comment by GuillaumeGomez** [comment]
I think it's a good idea to show only the linked theme in case "sync with system" is used. I'll update the PR.

**Comment by GuillaumeGomez** [comment]
I updated the PR and the online version. So I took the idea of having the "preferred settings" when selecting "system preference". It's much more intuitive this way in my opinion. So what do you think?

**Comment by GuillaumeGomez** [comment]
cc @steffahn and @apiraino :)

**Comment by steffahn** [comment]
Certainly an improvement over the status quo; of course I’d still prefer to also have the things that I suggested to remove more visual clutter (and hide options with no immediately visible effects) by:
* categorizing themes as “dark” vs. “light”, and only showing options that match this categorization in the “preferred dark theme”/“preferred light theme” options
  * consequently, if the “preferred light theme” category ends up with only a single option, then it’s *always* hidden
* only showing *either* of “preferred dark theme”/“preferred light theme”, depending on what the *current* system setting is (while of course still *saving* the other option, so that users where the system setting changes with day/night cycle only have to configure each of their day-theme an night-theme once).

and I also wouldn’t mind if the “preferred dark theme”/“preferred light theme” sections were more clearly labeled to be related to the system settings property.

The first thing, categorizing the themes as dark/light, would probably somehow need to interact with custom themes. Presumably it’s not a terrible idea to allow for categorizing a theme as “both dark and light”, too (so such a theme would still appear in *both* of the “preferred … theme” settings), and this could be a reasonable default for custom themes that didn’t specify anything else (yet), or even as a preliminary implementation strategy where the property of being light or dark could be hard-wired for the 3 standard themes, in case it isn’t clear yet what’s the best way for custom themes to specify this property.

---

Of course all of these suggested improvements are things one could also do as a subsequent change.

**Comment by GuillaumeGomez** [comment]
> categorizing themes as “dark” vs. “light”, and only showing options that match this categorization in the “preferred dark theme”/“preferred light theme” options 
> ...

I'm not a big fan of this. For example, someone could want to use "dark" on light system theme and "ayu" on dark system theme. So if we only display the "light" or "dark" themes, we prevent them from making this choice. And like you said below, it would need to interact with user added themes, which adds more complexity. I don't think it's worth it.

> and I also wouldn’t mind if the “preferred dark theme”/“preferred light theme” sections were more clearly labeled to be related to the system settings property.

I was thinking the same but apart from adding some left margin, not much ideas. If you have some, please share with us! :)

**Comment by steffahn** [comment]
Regarding the implementation, there might be benefit in keeping the representation in local storage exactly the same as before, since it isn’t uncommon to be switching between different versions of rustdoc pages, e.g. throughout `docs.rs`, or by looking at older versions of std docs with the `doc.rust-lang.org/…version…/std/` urls, so introducing problems in compatibility should be avoided; that is, assuming that this PR even changes how the configuration is stored in a breaking way… I haven’t tested that.

**Comment by steffahn** [comment]
> > and I also wouldn’t mind if the “preferred dark theme”/“preferred light theme” sections were more clearly labeled to be related to the system settings property.
> 
> I was thinking the same but apart from adding some left margin, not much ideas. If you have some, please share with us! :)

I did have a suggestion in [this comment](https://github.com/rust-lang/rust/pull/98765#issuecomment-1172947485).

**Comment by GuillaumeGomez** [comment]
> > > and I also wouldn’t mind if the “preferred dark theme”/“preferred light theme” sections were more clearly labeled to be related to the system settings property.
> > 
> > 
> > I was thinking the same but apart from adding some left margin, not much ideas. If you have some, please share with us! :)
> 
> I did have a suggestion in [this comment](https://github.com/rust-lang/rust/pull/98765#issuecomment-1172947485).

Saw it but not a big fan as it takes a lot of space and isn't great for visual "link" between the two.

> Regarding the implementation, there might be benefit in keeping the representation in local storage exactly the same as before, since it isn’t uncommon to be switching between different versions of rustdoc pages, e.g. throughout `docs.rs`, or by looking at older versions of std docs with the `doc.rust-lang.org/…version…/std/` urls, so introducing problems in compatibility should be avoided; that is, assuming that this PR even changes how the configuration is stored in a breaking way… I haven’t tested that.

I thought about it, but not sure how useful it would be. Now, the setting is much simpler to handle since it's handled directly into `rustdoc-theme`. Maybe I can simply check if `rustdoc-use-system-theme` is there and take its value before removing it.

**Comment by steffahn** [comment]
> > categorizing themes as “dark” vs. “light”, and only showing options that match this categorization in the “preferred dark theme”/“preferred light theme” options
> > ...
> 
> I'm not a big fan of this. For example, someone could want to use "dark" on light system theme and "ayu" on dark system theme. So if we only display the "light" or "dark" themes, we prevent them from making this choice. And like you said below, it would need to interact with user added themes, which adds more complexity. I don't think it's worth it.

There’s also the idea to have a (small) “advanced “system preference” customization” toggle/option which un-hides
* the currently inactive section
* the light themes in “preferred dark theme” and vice-versa
* (and of course also unhides the sections that were previously hidden because there was only a single visible choice)

A follow-up question would arise what happens if you disable the “advanced “system preference” customization” while your active setting is in a state that isn’t achievable without them (i.e. you have “preferred dark theme” that’s considered “light” or something like that); either prevent collapsing the advanced settings entirely, or *change* the settings to something non-advanced, or adjust the display rules (of the non-“advanced” view of the settings) so that e.g. in “preferred light theme” not only the light themes are displayed, but also the currently selected one.

**Comment by steffahn** [comment]
> I thought about it, but not sure how useful it would be. Now, the setting is much simpler to handle since it's handled directly into `rustdoc-theme`. Maybe I can simply check if `rustdoc-use-system-theme` is there and take its value before removing it.

Just saying, I can attest that it’s *still* annoying enough that viewing the documentation on `docs.rs` for much older crates sometimes changes themes because dark theme (or ayu) isn’t available on those pages yet. I feel like ensuring to make the user-experience better in this regard can be ***much more*** valuable than making the internal way of handling settings “simpler”, especially in case the settings still encode essentially the same set of possible states. (I’m not 100% sure how much this is the case actually, I’d still have to understand how exactly these settings work in the first place, and how it changed in this PR.)

While you *can* add “migration-code” for converting old-style data into the new format, the other way presumably can’t work.

**Comment by GuillaumeGomez** [comment]
> While you can add “migration-code” for converting old-style data into the new format, the other way presumably can’t work.

What other way? You mean going back to an older version?

Well in any case: if you have "use system theme" to `true`, "rustdoc-theme" isn't set. If "use system theme" is set to `false`, we use "rustdoc-theme". So in our case, if "rustdoc-theme" is empty, we default to "system". So basically, there is nothing to be done to keep it working as currently.

**Comment by steffahn** [comment]
> You mean going back to an older version?

Yes.

---

Maybe, instead of me trying to understand the whole thing *in theory*, I will just test it. E.g. switching back and forth between

https://rustdoc.crud.net/imperio/simplify-theme-settings/foo/index.html

and

https://rustdoc.crud.net/imperio/improve_rustdoc_search_results_page_crates_selection/foo/index.html

should probably be representative of the behavior we’re going to see in practice. I might not have enough time to test all corner cases until later today, I’ll report back on what problems I do or do not observe.

**Comment by steffahn** [comment]
> But then you increase the visual complexity once again, which is the opposite of what we're trying to here. wink

Not necessarily, I believe; especially if you account for the complexity that’s removed (by default) by such a change, and if the new toggle is visually simple and small, perhaps just a “show more” button in the right place.

As stated above, I do *not* suggest that any of this *needs to* be part of this same PR though.[^1] It could fit nicely in a separate PR as a “hide some stuff by default with an option to un-hide it if required” feature that only *adds* a visual-only enhancement feature, i.e. this wouldn’t entail any non-visual changes, it’s only about hiding some uncommon options in a smart manner; maybe I’ll give it a try myself. Even later, the option to categorize custom themes could then be yet-another incremental improvement (as that’s going to be slightly more than a “visual-only” change, since there’ll need to be a way of actually configuring this stuff for custom themes).

[^1]: I.e. ***the only*** concern I currently have with this PR is the problem of breaking the local storage data for possibly no good reason.

**Comment by GuillaumeGomez** [comment]
> If I start out on a fresh incognito-tab session with the after-this-PR page, then the settings start out in a weird state looking like

Good catch, it was a bug: I forgot to put back a `default` value.

About supporting both versions at the same time: the gain is minimal (switching a theme is a matter of seconds) whereas the maintenance cost would be quite high. The code allow to go from a previous version to this new one without problem. Supporting both ways (and until when?) doesn't seem worth it to me.

**Comment by steffahn** [comment]
> The code allow to go from a previous version to this new one without problem.

What do you mean by that?

> Supporting both ways (and until when?) doesn't seem worth it to me.

Certainly, keeping these settings the same forever isn’t necessary, and would be too limiting. However, I’d appreciate that *if* it changes in a breaking way, then
1. it stays possible to configure *both versions*, potentially independently, to “the same” setting that you like
2. merely opening an older-version rustdoc doesn’t change the config for the new-version, and vice-versa
3. to the degree that it’s easily possible, if you configure your theme in a new version rustdoc-generated content, then it also sets options so that the older versions show the right theme when opened, if that theme is supported

The desired behavior that’s a consequence of the above would be
* if you keep switching between older and newer rustdoc ***(as is super common on `docs.rs`)***, then you should be able to, eventually, reach a state where you don’t need to keep changing theme settings
* merely opening a different website shouldn’t mess with any of your settings
* if you occasionally land on a very old rustdoc page in the future, it would be *very nice* if the correct theme (if it’s even supported) was already selected

---

For vastly breaking changes in how the settings are stored, a straightforward approach would be to make sure that *different (new) keys* are used in order to not mess with any existing configuration for older versions of rustdoc on the same domain. (This would already ensure the properties 1. and 2. above.)

When the settings *are changed* it would be nice if the current settings state was translated into the older format, too, provided that it’s easily possible to store the current settings state in the old format either exactly or approximately. By making sure this only happens *when the settings are changed*, you still allow users to modify the translated settings on an old-version rustdoc page afterwards, without those changes then being overwritten immediately once you open a newer rustdoc page again.

(As an additional point, for migration purposes, it would also be *nice* if existing old-style settings were imported when new-version settings do not exist yet.)

---

For mostly cosmetic changes, it would be the most straightforward approach to not modify the format of the stored data at all.

---

The change as is, violates many of the desired behavior listed above:
* opening an old rustdoc page ***destroys*** your system-preference setting immediately
  * so it’s neither possible to configure both versions to the desired (matching) setting (if you desire a “system preference” setting),
  * nor is it ruled out that “merely opening a different website shouldn’t mess with any of your settings”

and the bonus property
* an existing “system preference” configuration doesn’t translate to the new version; even worse *the default configuration* (being “system preference”) does not translate to *the default configuration* (also being “system preference”) in the new version!

---

TL;DR, even if you don’t agree with any more general principles I’ve suggested above…
* being okay with the fact that merely looking at some older rustdoc page on the same domain will immediately (and somewhat invisibly) disable your custom “system preference” settings, and
* making sure not to break somewhat niche cases like “someone could want to use "dark" on light system theme and "ayu" on dark system theme”

don’t really fit into the same picture IMO.

---

In particular the first part of the 2nd property above, “merely opening an older-version rustdoc doesn’t change the config for the new-version”, seems absolutely essential to me, and it’s still broken in two ways:
* opening an older-version rustdoc page (say in a new tab you never even looked at yet) will *always* kick you out of “system preference”
* if you (ever touched system preference settings and then) go to a theme that does *not* match the current preferred … theme, then opening an older-version rustdoc page will change your theme

These are the two cases I showed in more details in the screenshots above. (For the second one example, you can add the lasst step;: if you open the new-style page at the end, it will have changed from “ayu” to “dark”, too.)

**Comment by steffahn** [comment]
Here’s another bug: Navigate to https://rustdoc.crud.net/imperio/simplify-theme-settings/foo/index.html in a new incognito session, where your system is configured dark (so system settings, the default, shows dark theme), then navigate to some other website, then press the back button. This leaves me in a light theme until I refresh the page or start playing around with theme options.

**Comment by GuillaumeGomez** [comment]
> > The code allow to go from a previous version to this new one without problem.
> 
> What do you mean by that?

What I wrote previously:

> Well in any case: if you have "use system theme" to `true`, "rustdoc-theme" isn't set. If "use system theme" is set to `false`, we use "rustdoc-theme". So in our case, if "rustdoc-theme" is empty, we default to "system". So basically, there is nothing to be done to keep it working as currently.

> if you keep switching between older and newer rustdoc (as is super common on docs.rs), then you should be able to, eventually, reach a state where you don’t need to keep changing theme settings

There is one case that doesn't work: if you select "system theme" on the new version.

> merely opening a different website shouldn’t mess with any of your settings

I don't understand this comment. It's stored in local storage, so nothing to do with different websites.

> if you occasionally land on a very old rustdoc page in the future, it would be very nice if the correct theme (if it’s even supported) was already selected

Same answer as above.

> When the settings are changed it would be nice if the current settings state was translated into the older format, too, provided that it’s easily possible to store the current settings state in the old format either exactly or approximately. By making sure this only happens when the settings are changed, you still allow users to modify the translated settings on an old-version rustdoc page afterwards, without those changes then being overwritten immediately once you open a newer rustdoc page again.

This is exactly what I told you: it's literally two clicks to change it. Maximum a few seconds. I don't think it's worth adding **more** JS code to handle that. Please keep in mind that we aren't many working on this and that our time is **very** limited.


> The change as is, violates many of the desired behavior listed above:
> 
> * opening an old rustdoc page _**destroys**_ your system-preference setting immediately
>   
>   * so it’s neither possible to configure both versions to the desired (matching) setting (if you desire a “system preference” setting),
>   * nor is it ruled out that “merely opening a different website shouldn’t mess with any of your settings”

Supporting cross-version doesn't really make sense. And again, I don't understand what you're trying to say about "opening another website shouldn't mess any of your settings". It's website-based.

> * being okay with the fact that merely looking at some older rustdoc page on the same domain will immediately (and somewhat invisibly) disable your custom “system preference” settings, and
> * making sure not to break somewhat niche cases like “someone could want to use "dark" on light system theme and "ayu" on dark system theme”
> 
> don’t really fit into the same picture IMO.

They do: we want to keep to allow users to be able to pick their "system theme" preference while improving the UI/UX. It's very possible to write a script on `docs.rs` that could handle this corner case (not sure if it's something that we'd want though, but still possible).

> Here’s another bug: Navigate to https://rustdoc.crud.net/imperio/simplify-theme-settings/foo/index.html in a new incognito session, where your system is configured dark (so “system preference” settings, the default, show dark theme), then navigate to some other website, then press the back button. This leaves me (testing in Google Chrome) in a light theme until I refresh the page or start playing around with theme options.
> 
> The behavior disappears once I change “Theme” to anything other than “system preference” and then back to “system preference”, which suggests that this is also some kind of problem with the default values.

Thanks for the issue! I'll try to take a look.

**Comment by steffahn** [comment]
> > merely opening a different website shouldn’t mess with any of your settings
> 
> I don't understand this comment. It's stored in local storage, so nothing to do with different websites.

Sorry if that wasn’t clear: “different website”, but same domain. I’ve mentioned my main example: docs.rs. It’ll take months, probably years, until you’ll stop regularly coming across documentation pages of older crates that won’t yet have docs from after this PR.


> > When the settings are changed it would be nice if the current settings state was translated into the older format, too, provided that it’s easily possible to store the current settings state in the old format either exactly or approximately. By making sure this only happens when the settings are changed, you still allow users to modify the translated settings on an old-version rustdoc page afterwards, without those changes then being overwritten immediately once you open a newer rustdoc page again.
> 
> This is exactly what I told you: it's literally two clicks to change it. Maximum a few seconds. I don't think it's worth adding **more** JS code to handle that. Please keep in mind that we aren't many working on this and that our time is **very** limited.

This is just an “it would be nice”, so nothing I care about a lot.

My main point is that the data format should, in my opinion (as far as approaches I could come up with go) either stay unmodified, or for breaking changes, choosing a new (essentially versioned) set of local storage keys so that configurations for the previous version and the new version can coexist. I’ve just learned that inspecting local storage in the browse is apparently fairly straighforward, so I myself can try to understand the concrete details of what’s going on / changing with this PR and maybe give a *concrete* suggestion what sensibly might be changed.

---

Maybe something easy like changing a single key name, in particular, using “`rustdoc-theme-2`” (or something like “`rustdoc-theme-or-system`”) instead of “`rustdoc-theme`” – the latter can then keep the old, unmodified, value for older-version rustdoc under the same domain – is enough to avoid the problems I’ve been experiencing, that visiting an older rustdoc website (under the same domain) will modify my settings. Even without any migration logic in either direction that could probably result in a significant improvement in user experience (as you mentioned, configuring the theme is straightforward; without migration logic, users will have to do it twice, but no more than that). And it wouldn’t cost any “adding more JS code” or significant development time. Realistically, adding any extra (migration) logic would most likely end up with a more complex result than if the new UI was made to work with the old-style local storage data anyways, so if absolute simplicity of the JS code is really important (i.e. an argument against keeping the local storage data format unmodified), then renaming the key like that could work out fairly nicely.

**Comment by GuillaumeGomez** [comment]
Renaming the setting would do the trick, indeed!

**Comment by GuillaumeGomez** [comment]
I renamed `rustdoc-theme` to rustdoc-theme2` as suggested. To make things simpler, I grouped the calls to the `theme` settings into a function which looks from the previous version.

The last commit is some dead code that was just laying so I removed it.

Does it look good to you @steffahn ?

**Comment by GuillaumeGomez** [comment]
Fixed eslint error.

**Comment by steffahn** [comment]
> Does it look good to you @steffahn ?

Sorry for not answering more quickly. Taking a look now; did you update the uploaded code to https://rustdoc.crud.net/imperio/simplify-theme-settings/foo/index.html ? If yes, note that it still seems to be setting `rustdoc-theme`, not `rustdoc-theme2` when I modify the Theme settings. If no, I would appreciate an update so I can properly re-do the tests cases I listed in previous comments.

---

Upon further looking it looks like you *did* update it, so there’s apparently still a problem.

Also I question the usefulness of the migration logic as-is, because reading from an existing `rustdoc-theme` value if `rustdoc-theme2` doesn’t exist yet means that a `system-preference` setting wouldn’t be translated properly. This seems problematic since it’s the *default option*, and `rustdoc-theme` on current rustdoc pages will be automatically set to the current theme without the user touching the options; i.e. opening an unconfigured/default-setting old rustdoc page would influence opening a newer rustdoc page on the same domain later in a way that *it isn’t* in the unconfigured/default-setting state anymore. I’d personally prefer ***no*** translation logic for “try to get what's being used in the previous version” over a bad one; or, if you’d like to fix the translation logic, make sure to also check whether or not `"use-system-theme"` is true.

**Comment by steffahn** [comment]
So to fix the incomplete renaming, I suppose you’ll have to change the `name` property of the radio button to use `"theme2"` instead of `"theme"`.

File: src/librustdoc/html/static/js/settings.js [hunk 1]
```diff
@@ -149,16 +147,11 @@
...
         const themes = getVar("themes").split(",");
         const settings = [
-            {
-                "name": "Use system theme",
-                "js_name": "use-system-theme",
-                "default": true,
-            },
             {
                 "name": "Theme",
                 "js_name": "theme",
```
Comment (steffahn) [inline src/librustdoc/html/static/js/settings.js:152]: So I suppose here’s where it would still need to be renamed to `"theme2"`.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/settings.js:152]: Indeed, thanks!

**Comment by GuillaumeGomez** [comment]
> Also I question the usefulness of the migration logic as-is, because reading from an existing `rustdoc-theme` value if `rustdoc-theme2` doesn’t exist yet means that a `system-preference` setting wouldn’t be translated properly. This seems problematic since it’s the _default option_, and `rustdoc-theme` on current rustdoc pages will be automatically set to the current theme without the user touching the options; i.e. opening an unconfigured/default-setting old rustdoc page would influence opening a newer rustdoc page on the same domain later in a way that _it isn’t_ in the unconfigured/default-setting state anymore. I’d personally prefer _**no**_ translation logic for “try to get what's being used in the previous version” over a bad one; or, if you’d like to fix the translation logic, make sure to also check whether or not `"use-system-theme"` is true.

I was wondering whether it was worth it or not. Well, I'll implement it and see for myself how big it is.

**Comment by GuillaumeGomez** [comment]
I fixed the theme switching bug. I now also get the "use-system-theme" setting when there is no "rustdoc-theme2" set. I updated the online demo too.

**Comment by steffahn** [comment]
Still behaves weird in common cases. If I navigate to https://rustdoc.crud.net/imperio/improve_rustdoc_search_results_page_crates_selection/foo/index.html in an incognito tab, then switch to https://rustdoc.crud.net/imperio/simplify-theme-settings/foo/index.html then I get into a “system preference” state without the further “Preferred … theme” options visible.
Screenshot_20220728_181538

**Comment by GuillaumeGomez** [comment]
The settings were not checking the currently set values. Fixed it and added a test for it. Thanks a lot for checking this!

**Comment by GuillaumeGomez** [comment]
Did you have time to take another look @steffahn ?

**Comment by GuillaumeGomez** [comment]
Fixed conflict.

**Comment by GuillaumeGomez** [comment]
It's fine, don't worry. It's not really a hurry. Just wanted to be sure what was worrying you was fixed.

**Comment by GuillaumeGomez** [comment]
The UI change was merged in https://github.com/rust-lang/rust/pull/104366 so closing this one.
